File: leetcode/linked-list/445.两数相加-ii.js
/*
 * @lc app=leetcode.cn id=445 lang=javascript
 *
 * [445] 两数相加 II
 *
 * https://leetcode-cn.com/problems/add-two-numbers-ii/description/
 *
 * algorithms
 * Medium (45.69%)
 * Likes:    210
 * Dislikes: 0
 * Total Accepted:    40.4K
 * Total Submissions: 70.5K
 * Testcase Example:  '[7,2,4,3]\n[5,6,4]'
 *
 * 给你两个 非空 链表来代表两个非负整数。数字最高位位于链表开始位置。它们的每个节点只存储一位数字。将这两数相加会返回一个新的链表。
 *
 * 你可以假设除了数字 0 之外，这两个数字都不会以零开头。
 *
 *
 *
 * 进阶：
 *
 * 如果输入链表不能修改该如何处理？换句话说，你不能对列表中的节点进行翻转。
 *
 *
 *
 * 示例：
 *
 * 输入：(7 -> 2 -> 4 -> 3) + (5 -> 6 -> 4)
 * 输出：7 -> 8 -> 0 -> 7
 *
 *
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 * stack
 */
var addTwoNumbers = function (l1, l2) {
  const s1 = [];
  const s2 = [];

  while (l1 != null) {
    s1.push(l1.val);
    l1 = l1.next;
  }

  while (l2 != null) {
    s2.push(l2.val);
    l2 = l2.next;
  }

  let ret = null;
  let c = 0;

  while (s1.length || s2.length || c > 0) {
    let a = s1.length ? s1.pop() : 0;
    let b = s2.length ? s2.pop() : 0;
    let sum = a + b + c;
    let node = new ListNode(sum % 10);
    c = Math.floor(sum / 10);
    node.next = ret;
    ret = node;
  }

  return ret;
};
// @lc code=end

File: leetcode/linked-list/445.两数相加-ii.go
/*
 * @lc app=leetcode.cn id=445 lang=golang
 *
 * [445] 两数相加 II
 *
 * https://leetcode.cn/problems/add-two-numbers-ii/description/
 *
 * algorithms
 * Medium (60.21%)
 * Likes:    628
 * Dislikes: 0
 * Total Accepted:    128.5K
 * Total Submissions: 211.8K
 * Testcase Example:  '[7,2,4,3]\n[5,6,4]'
 *
 * 给你两个 非空 链表来代表两个非负整数。数字最高位位于链表开始位置。它们的每个节点只存储一位数字。将这两数相加会返回一个新的链表。
 *
 * 你可以假设除了数字 0 之外，这两个数字都不会以零开头。
 *
 *
 *
 * 示例1：
 *
 *
 *
 *
 * 输入：l1 = [7,2,4,3], l2 = [5,6,4]
 * 输出：[7,8,0,7]
 *
 *
 * 示例2：
 *
 *
 * 输入：l1 = [2,4,3], l2 = [5,6,4]
 * 输出：[8,0,7]
 *
 *
 * 示例3：
 *
 *
 * 输入：l1 = [0], l2 = [0]
 * 输出：[0]
 *
 *
 *
 *
 * 提示：
 *
 *
 * 链表的长度范围为 [1, 100]
 * 0 <= node.val <= 9
 * 输入数据保证链表代表的数字无前导 0
 *
 *
 *
 *
 * 进阶：如果输入链表不能翻转该如何解决？
 *
 */

package leetcode

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

// @lc code=start
func addTwoNumbers(l1 *ListNode, l2 *ListNode) (res *ListNode) {
	a1, a2 := []int{}, []int{}
	for l1 != nil {
		a1 = append(a1, l1.Val)
		l1 = l1.Next
	}
	for l2 != nil {
		a2 = append(a2, l2.Val)
		l2 = l2.Next
	}

	carry := 0
	for len(a1) > 0 || len(a2) > 0 || carry > 0 {
		sum := carry
		if len(a1) > 0 {
			sum += a1[len(a1)-1]
			a1 = a1[:len(a1)-1]
		}
		if len(a2) > 0 {
			sum += a2[len(a2)-1]
			a2 = a2[:len(a2)-1]
		}
		carry = sum / 10
		val := sum % 10
		node := &ListNode{Val: val}
		node.Next = res
		res = node
	}

	return
}

// @lc code=end
